firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4474,77 +4479,99 @@ public void getQueryBasic() throws DatabaseException, InterruptedException, Exec
   }
 
   @Test
-  public void getQueryCached()
-      throws DatabaseException, InterruptedException, TimeoutException, TestFailure,
-          ExecutionException {
-    FirebaseApp app =
-        appForDatabaseUrl(IntegrationTestValues.getAltNamespace(), UUID.randomUUID().toString());
-    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
-    DatabaseReference ref = db.getReference();
-    final Semaphore semaphore = new Semaphore(0);
-    ValueEventListener listener =
-        new ValueEventListener() {
-          @Override
-          public void onDataChange(@NonNull DataSnapshot snapshot) {
-            if (snapshot.getValue() != null && snapshot.getValue().equals(42L)) {
+  public void testGetProbesInMemoryCacheForActiveListenerWhenOffline()
+      throws DatabaseException, InterruptedException, ExecutionException, TestFailure,
+          TimeoutException {
+    FirebaseApp readerApp =
+        appForDatabaseUrl(IntegrationTestValues.getNamespace(), UUID.randomUUID().toString());
+    FirebaseApp writerApp =
+        appForDatabaseUrl(IntegrationTestValues.getNamespace(), UUID.randomUUID().toString());
+    FirebaseDatabase readerDb = FirebaseDatabase.getInstance(readerApp);
+    readerDb.setPersistenceEnabled(true);
+    FirebaseDatabase writerDb = FirebaseDatabase.getInstance(writerApp);
+
+    DatabaseReference reader = readerDb.getReference();
+    DatabaseReference writer = writerDb.getReference();
+    String key = reader.push().getKey();
+    reader = reader.child(key);
+    writer = writer.child(key);
+
+    try {
+      new WriteFuture(writer, 42L).timedGet();
+
+      assertEquals(Tasks.await(reader.get()).getValue(), 42L);
+
+      Semaphore semaphore = new Semaphore(0);
+      reader.addValueEventListener(
+          new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+              assertEquals(snapshot.getValue(), 42L);
               semaphore.release();
             }
-          }
 
-          @Override
-          public void onCancelled(@NonNull DatabaseError error) {}
-        };
-    ref.addValueEventListener(listener);
-    ref.setValue(42L);
-    IntegrationTestHelpers.waitFor(semaphore);
-    db.goOffline();
-    try {
-      // Since we still have a listener on `ref`, the 42L should be cached here.
-      assertEquals(42L, Tasks.await(ref.get()).getValue());
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {
+              fail(error.getMessage());
```

Comment:
Note that this won't actually fail the test since it is running in a different thread. It will however print the error to the console, and the test should fail because the semaphore times out.